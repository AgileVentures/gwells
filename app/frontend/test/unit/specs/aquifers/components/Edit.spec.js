import { shallowMount, createLocalVue } from '@vue/test-utils'
import EditComponent from '@/aquifers/components/Edit.vue'
import Vuex from 'vuex'
import axios from 'axios'
import auth from '@/common/store/auth.js'

jest.mock('axios')

const localVue = createLocalVue()
localVue.use(Vuex)

const aquiferFixture = {
  aquifer_id: '4',
  aquifer_name: 'Aggasiz',
  location_description: 'Aggasiz/Sea Bird Island',
  material_description: 'Sand and Gravel',
  litho_stratographic_unit: 'Fraser River Sediments',
  subtype_description: 'Unconfined sand and gravel - large river system',
  vulnerability_description: 'High',
  area: '63.3',
  productivity_description: 'High',
  demand: 'Low',
  mapping_year: '1993'
}

describe('Edit Component', () => {
  const component = (options) => shallowMount(EditComponent, {
    localVue,
    data() {
      return {
        record: aquiferFixture
      }
    },
    store: new Vuex.Store({
      modules: { auth }
    }),
    methods: {
      fetchCode() { },
      fetch() { }
    },
    propsData: {
      id: 10
    },
    ...options
  })

  it('Matches the snapshot', () => {
    const wrapper = component()

    expect(wrapper.element).toMatchSnapshot()
  })

  it('queries aquifer on load', () => {
    const fetch = jest.fn()

    component({ methods: { fetchCode() { }, fetch } })

    expect(fetch).toHaveBeenCalled()
  })

  it('displays field errors messages', () => {
    let errorMessage = 'error message'

    const wrapper = component({
      data() {
        return {
          fieldErrors: {
            mapping_year: [errorMessage],
            aquifer_name: [errorMessage],
            litho_stratographic_unit: [errorMessage],
            location_description: [errorMessage],
            vulnerability: [errorMessage],
            material: [errorMessage],
            subtype: [errorMessage],
            quality_concerns: [errorMessage],
            productivity: [errorMessage],
            area: [errorMessage],
            demand: [errorMessage],
            known_water_use: [errorMessage],
            notes: [errorMessage],
          }
        }
      }
    })

    expect(wrapper.find('#mapping_year + [role="alert"]').text()).toBe(errorMessage)
    expect(wrapper.find('#aquifer_name + [role="alert"]').text()).toBe(errorMessage)
    expect(wrapper.find('#litho_stratographic_unit + [role="alert"]').text()).toBe(errorMessage)
    expect(wrapper.find('#location_description + [role="alert"]').text()).toBe(errorMessage)
    expect(wrapper.find('#vulnerability + [role="alert"]').text()).toBe(errorMessage)
    expect(wrapper.find('#material + [role="alert"]').text()).toBe(errorMessage)
    expect(wrapper.find('#subtype + [role="alert"]').text()).toBe(errorMessage)
    expect(wrapper.find('#quality_concerns + [role="alert"]').text()).toBe(errorMessage)
    expect(wrapper.find('#productivity + [role="alert"]').text()).toBe(errorMessage)
    expect(wrapper.find('#area + [role="alert"]').text()).toBe(errorMessage)
    expect(wrapper.find('#demand + [role="alert"]').text()).toBe(errorMessage)
    expect(wrapper.find('#known_water_use + [role="alert"]').text()).toBe(errorMessage)
    expect(wrapper.find('#notes + [role="alert"]').text()).toBe(errorMessage)
  })

  describe('save', () => {
    it('sends a patch with the contents of record on save', () => {
      const wrapper = component()

      axios.patch.mockResolvedValue(true)

      wrapper.vm.save()

      expect(axios.patch).toHaveBeenCalledWith('aquifers/10/', aquiferFixture)
    })
  })
})
