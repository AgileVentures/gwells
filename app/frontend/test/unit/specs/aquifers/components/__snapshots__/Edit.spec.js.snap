// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Edit Component Matches the snapshot 1`] = `
<div>
  <div
    class="card p-3 mb-4 container container-wide"
  >
    <!---->
     
    <!---->
     
    <div
      class="container"
    >
      <div
        class="row border-bottom mb-3 pb-2"
      >
        <div
          class="col"
        >
          <h1>
            Aquifer 4 Summary - Edit
          </h1>
        </div>
      </div>
       
      <!---->
       
      <div
        class="row"
      >
        <div
          class="col-md-6"
        >
          <div
            class="row"
          >
            <div
              class="col-form-label col-md-4"
            >
              
              Aquifer number
            
            </div>
             
            <div
              class="col-form-label col-md-8"
            >
              
              4
            
            </div>
          </div>
        </div>
         
        <div
          class="col-md-6"
        >
          <div
            aria-labelledby="__BVID__5__BV_label_"
            class="b-form-group form-group"
            id="__BVID__5"
            role="group"
          >
            <div
              class="form-row"
            >
              <label
                class="col-sm-4 col-form-label"
                for="mapping_year"
                id="__BVID__5__BV_label_"
              >
                Year of mapping
              </label>
              <div
                class="col-sm-8"
              >
                <input
                  aria-required="true"
                  class="form-control"
                  id="mapping_year"
                  required="required"
                  type="text"
                  value="1993"
                />
                <!---->
                <!---->
                <!---->
              </div>
            </div>
          </div>
        </div>
      </div>
       
      <div
        class="row"
      >
        <div
          class="col-md-6"
        >
          <div
            aria-labelledby="__BVID__7__BV_label_"
            class="b-form-group form-group"
            id="__BVID__7"
            role="group"
          >
            <div
              class="form-row"
            >
              <label
                class="col-sm-4 col-form-label"
                for="aquifer_name"
                id="__BVID__7__BV_label_"
              >
                Aquifer name
              </label>
              <div
                class="col-sm-8"
              >
                <input
                  aria-required="true"
                  class="form-control"
                  id="aquifer_name"
                  required="required"
                  type="text"
                  value="Aggasiz"
                />
                <!---->
                <!---->
                <!---->
              </div>
            </div>
          </div>
        </div>
         
        <div
          class="col-md-6"
        >
          <div
            aria-labelledby="__BVID__9__BV_label_"
            class="b-form-group form-group"
            id="__BVID__9"
            role="group"
          >
            <div
              class="form-row"
            >
              <label
                class="col-sm-4 col-form-label"
                for="litho_stratographic_unit"
                id="__BVID__9__BV_label_"
              >
                Litho stratigraphic unit
              </label>
              <div
                class="col-sm-8"
              >
                <input
                  aria-required="true"
                  class="form-control"
                  id="litho_stratographic_unit"
                  required="required"
                  type="text"
                  value="Fraser River Sediments"
                />
                <!---->
                <!---->
                <!---->
              </div>
            </div>
          </div>
        </div>
      </div>
       
      <div
        class="row"
      >
        <div
          class="col-md-6"
        >
          <div
            aria-labelledby="__BVID__11__BV_label_"
            class="b-form-group form-group"
            id="__BVID__11"
            role="group"
          >
            <div
              class="form-row"
            >
              <label
                class="col-sm-4 col-form-label"
                for="location_description"
                id="__BVID__11__BV_label_"
              >
                Descriptive location
              </label>
              <div
                class="col-sm-8"
              >
                <input
                  aria-required="true"
                  class="form-control"
                  id="location_description"
                  required="required"
                  type="text"
                  value="Aggasiz/Sea Bird Island"
                />
                <!---->
                <!---->
                <!---->
              </div>
            </div>
          </div>
        </div>
         
        <div
          class="col-md-6"
        >
          <div
            aria-labelledby="__BVID__13__BV_label_"
            class="b-form-group form-group"
            id="__BVID__13"
            role="group"
          >
            <div
              class="form-row"
            >
              <label
                class="col-sm-4 col-form-label"
                for="vulnerability"
                id="__BVID__13__BV_label_"
              >
                Vulnerability
              </label>
              <div
                class="col-sm-8"
              >
                <select
                  aria-required="true"
                  class="form-control custom-select"
                  id="vulnerability"
                  required="required"
                />
                <!---->
                <!---->
                <!---->
              </div>
            </div>
          </div>
        </div>
      </div>
       
      <div
        class="row"
      >
        <div
          class="col-md-6"
        >
          <div
            aria-labelledby="__BVID__15__BV_label_"
            class="b-form-group form-group"
            id="__BVID__15"
            role="group"
          >
            <div
              class="form-row"
            >
              <label
                class="col-sm-4 col-form-label"
                for="material"
                id="__BVID__15__BV_label_"
              >
                Material type
              </label>
              <div
                class="col-sm-8"
              >
                <select
                  aria-required="true"
                  class="form-control custom-select"
                  id="material"
                  required="required"
                />
                <!---->
                <!---->
                <!---->
              </div>
            </div>
          </div>
        </div>
         
        <div
          class="col-md-6"
        >
          <div
            aria-labelledby="__BVID__17__BV_label_"
            class="b-form-group form-group"
            id="__BVID__17"
            role="group"
          >
            <div
              class="form-row"
            >
              <label
                class="col-sm-4 col-form-label"
                for="subtype"
                id="__BVID__17__BV_label_"
              >
                Subtype
              </label>
              <div
                class="col-sm-8"
              >
                <select
                  aria-required="true"
                  class="form-control custom-select"
                  id="subtype"
                  required="required"
                />
                <!---->
                <!---->
                <!---->
              </div>
            </div>
          </div>
        </div>
      </div>
       
      <div
        class="row"
      >
        <div
          class="col-md-6"
        >
          <div
            aria-labelledby="__BVID__19__BV_label_"
            class="b-form-group form-group"
            id="__BVID__19"
            role="group"
          >
            <div
              class="form-row"
            >
              <label
                class="col-sm-4 col-form-label"
                for="quality_concern"
                id="__BVID__19__BV_label_"
              >
                Quality concerns
              </label>
              <div
                class="col-sm-8"
              >
                <select
                  aria-required="true"
                  class="form-control custom-select"
                  id="quality_concern"
                  required="required"
                />
                <!---->
                <!---->
                <!---->
              </div>
            </div>
          </div>
        </div>
         
        <div
          class="col-md-6"
        >
          <div
            aria-labelledby="__BVID__21__BV_label_"
            class="b-form-group form-group"
            id="__BVID__21"
            role="group"
          >
            <div
              class="form-row"
            >
              <label
                class="col-sm-4 col-form-label"
                for="productivity"
                id="__BVID__21__BV_label_"
              >
                Productivity
              </label>
              <div
                class="col-sm-8"
              >
                <select
                  aria-required="true"
                  class="form-control custom-select"
                  id="productivity"
                  required="required"
                />
                <!---->
                <!---->
                <!---->
              </div>
            </div>
          </div>
        </div>
      </div>
       
      <div
        class="row"
      >
        <div
          class="col-md-6"
        >
          <div
            aria-labelledby="__BVID__23__BV_label_"
            class="b-form-group form-group"
            id="__BVID__23"
            role="group"
          >
            <div
              class="form-row"
            >
              <label
                class="col-sm-4 col-form-label"
                for="area"
                id="__BVID__23__BV_label_"
              >
                Size (km²)
              </label>
              <div
                class="col-sm-8"
              >
                <input
                  aria-required="true"
                  class="form-control"
                  id="area"
                  required="required"
                  type="text"
                  value="63.3"
                />
                <!---->
                <!---->
                <!---->
              </div>
            </div>
          </div>
        </div>
         
        <div
          class="col-md-6"
        >
          <div
            aria-labelledby="__BVID__25__BV_label_"
            class="b-form-group form-group"
            id="__BVID__25"
            role="group"
          >
            <div
              class="form-row"
            >
              <label
                class="col-sm-4 col-form-label"
                for="demand"
                id="__BVID__25__BV_label_"
              >
                Demand
              </label>
              <div
                class="col-sm-8"
              >
                <select
                  aria-required="true"
                  class="form-control custom-select"
                  id="demand"
                  required="required"
                />
                <!---->
                <!---->
                <!---->
              </div>
            </div>
          </div>
        </div>
      </div>
       
      <div
        class="row"
      >
        <div
          class="col-md-6"
        >
          <div
            aria-labelledby="__BVID__27__BV_label_"
            class="b-form-group form-group"
            id="__BVID__27"
            role="group"
          >
            <div
              class="form-row"
            >
              <label
                class="col-sm-4 col-form-label"
                for="known_water_use"
                id="__BVID__27__BV_label_"
              >
                Type of known water use
              </label>
              <div
                class="col-sm-8"
              >
                <select
                  aria-required="true"
                  class="form-control custom-select"
                  id="known_water_use"
                  required="required"
                />
                <!---->
                <!---->
                <!---->
              </div>
            </div>
          </div>
        </div>
      </div>
       
      <div
        class="row"
      >
        <div
          class="col-md-6"
        >
          <div
            aria-labelledby="__BVID__29__BV_label_"
            class="b-form-group form-group"
            id="__BVID__29"
            role="group"
          >
            <div
              class="form-row"
            >
              <label
                class="col-sm-4 col-form-label"
                for="notes"
                id="__BVID__29__BV_label_"
              >
                Comments
              </label>
              <div
                class="col-sm-8"
              >
                <textarea
                  class="form-control"
                  id="notes"
                  rows="4"
                  wrap="soft"
                />
                <!---->
                <!---->
                <!---->
              </div>
            </div>
          </div>
        </div>
      </div>
       
      <div
        class="row mt-4"
      >
        <div
          class="col-auto"
        >
          <button
            class="btn btn-secondary"
            type="button"
          >
            
            Save
          
          </button>
           
          <button
            class="btn btn-secondary"
            type="button"
          >
            
            Cancel
          
          </button>
        </div>
      </div>
    </div>
  </div>
   
  <div>
    <div
      id="confirmSave___BV_modal_outer_"
    >
      <div
        aria-hidden="true"
        class="modal fade"
        id="confirmSave"
        role="dialog"
        style="display: none;"
      >
        <div
          class="modal-dialog modal-md"
        >
          <div
            aria-describedby="confirmSave___BV_modal_body_"
            aria-labelledby="confirmSave___BV_modal_header_"
            class="modal-content"
            role="document"
            tabindex="-1"
          >
            <header
              class="modal-header"
              id="confirmSave___BV_modal_header_"
            >
              <h5
                class="modal-title"
              />
              <button
                aria-label="Close"
                class="close"
                type="button"
              >
                ×
              </button>
            </header>
            <div
              class="modal-body"
              id="confirmSave___BV_modal_body_"
            >
              <p>
                Are you sure you would like to save this record?
              </p>
            </div>
            <footer
              class="modal-footer"
              id="confirmSave___BV_modal_footer_"
            >
              <button
                class="btn btn-secondary"
                type="button"
              >
                Cancel
              </button>
              <button
                class="btn btn-primary"
                type="button"
              >
                OK
              </button>
            </footer>
          </div>
        </div>
      </div>
      <!---->
    </div>
  </div>
   
  <div>
    <div
      id="confirmCancel___BV_modal_outer_"
    >
      <div
        aria-hidden="true"
        class="modal fade"
        id="confirmCancel"
        role="dialog"
        style="display: none;"
      >
        <div
          class="modal-dialog modal-md"
        >
          <div
            aria-describedby="confirmCancel___BV_modal_body_"
            aria-labelledby="confirmCancel___BV_modal_header_"
            class="modal-content"
            role="document"
            tabindex="-1"
          >
            <header
              class="modal-header"
              id="confirmCancel___BV_modal_header_"
            >
              <h5
                class="modal-title"
              />
              <button
                aria-label="Close"
                class="close"
                type="button"
              >
                ×
              </button>
            </header>
            <div
              class="modal-body"
              id="confirmCancel___BV_modal_body_"
            >
              <p>
                Are you sure you want to quit editing this record?
              </p>
            </div>
            <footer
              class="modal-footer"
              id="confirmCancel___BV_modal_footer_"
            >
              <button
                class="btn btn-secondary"
                type="button"
              >
                Cancel
              </button>
              <button
                class="btn btn-primary"
                type="button"
              >
                OK
              </button>
            </footer>
          </div>
        </div>
      </div>
      <!---->
    </div>
  </div>
</div>
`;
